File: 0721. Accounts Merge.java
/*
721. Accounts Merge
union-find로 풀어야 하는데 map을 써야 한다.
map<email, account> account 안에 set<email>, name, parent를 넣는다.

일단 각 Account 별로 account 객체를 만든다.
그리고 그것의 email을 하나씩 돌면서 map에 key=email이 있는지 찾아서, 있으면 union한다. 그리고 map.put() 한다.

이렇게 돌고 나면 Account들이 다 합쳐져 있을 것이다.
문제는 map의 key가 아니라 value로 들어있는 거라, map을 돌면서 distinct account를 해야 한다. 이건 set을 쓰면 된다.
emails.sort()도 필요하다.
*/
class Solution {
    public List<List<String>> accountsMerge(List<List<String>> accounts) {
        Map<String, Account> map = new HashMap<>();
        for (List<String> account : accounts) {
            Account a = new Account(account.get(0));
            for (int i = 1; i < account.size(); i++) {
                String email = account.get(i);
                a.emails.add(email);
                if (map.containsKey(email)) a = a.union(map.get(email).find());
                map.put(email, a);
            }
        }
        List<List<String>> res = new ArrayList<>();
        Set<Account> set = new HashSet<>();
        for (Account a : map.values()) {
            Account p = a.find();
            if (set.contains(p)) continue;
            set.add(p);
            List<String> account = new ArrayList<>();
            account.add(p.name);
            List<String> emails = new ArrayList<>(p.emails);
            Collections.sort(emails);
            account.addAll(emails);
            res.add(account);
        }
        return res;
    }
    
}
class Account {
    String name;
    Set<String> emails;
    Account p;
    Account(String name) {
        this.name = name;
        emails = new HashSet<>();
        p = this;
    }
    Account union(Account a) {
        if (a == this) return a;
        p = a;
        p.emails.addAll(emails);
        return p;
    }
    Account find() {
        if (p == this) return p;
        return p = p.find();
    }
}

File: 0721. Accounts Merge.cs
/*
721. Accounts Merge
*/
// https://leetcode.com/problems/accounts-merge/description/
// n = length of all email addresses
// Time: O(nlogn) because of Union Find & List<>.Sort()
// Space: O(n)
public class Solution {
    private int findParent(int i, int[] accP){
        if(accP[i]!=i) accP[i] = findParent(accP[i], accP);
        return accP[i];
    }
    private void union(int i, int j, int[] accP, int[] accCnt){
        int iP = findParent(i, accP);
        int jP = findParent(j, accP);
        if(iP!=jP){
            if(accCnt[iP]>=accCnt[jP]){
                accP[jP] = iP;
                accCnt[iP] += accCnt[jP];
                accCnt[jP] = 0;
            }else{
                accP[iP] = jP;
                accCnt[jP] += accCnt[iP];
                accCnt[iP] = 0;
            }
        }
    }
    public IList<IList<string>> AccountsMerge(IList<IList<string>> accounts) {
        Dictionary<string,int> dic = new Dictionary<string,int>();
        string[] acc = new string[accounts.Count];
        int[] accCnt = new int[accounts.Count];
        int[] accP = new int[accounts.Count];
        int mergeCnt = 0;
        for(int i=0; i<accounts.Count; i++){
            acc[i] = accounts[i][0];
            accCnt[i] = accounts[i].Count-1;
            accP[i] = i;
            int accI = i;
            for(int j=1; j<accounts[i].Count; j++){
                if(dic.ContainsKey(accounts[i][j])){
                    union(dic[accounts[i][j]], accI, accP, accCnt);
                }else dic.Add(accounts[i][j], accI);
            }
        }
        int[] accMap = new int[acc.Length];
        int mapCnt = 0;
        IList<IList<string>> ansList = new List<IList<string>>();
        for(int i=0; i<accCnt.Length; i++){
            if(accCnt[i]>0){
                accMap[i] = mapCnt;
                ansList.Add(new List<string>());
                ansList[mapCnt].Add(acc[i]);
                mapCnt++;
            }
        }
        foreach(var kv in dic) ansList[accMap[findParent(kv.Value,accP)]].Add(kv.Key);
        for(int i=0; i<ansList.Count; i++){
            List<string> list = (List<string>)ansList[i];
            list.Sort(1, ansList[i].Count-1, StringComparer.Ordinal);
            ansList[i] = list;
        }        
        return ansList;
    }
}
